From 0f06bee297da04c21a8b0de2b81ccdb8a1923a08 Mon Sep 17 00:00:00 2001
From: Jiang Lu <lu.jiang@windriver.com>
Date: Fri, 16 Nov 2018 11:13:05 +0800
Subject: [PATCH] of: reserved_mem: disable kmemleak scan on removed memory
 blocks

Memory reserved with "nomap" DT property in of_reserved_mem.c
removes the memory block. The removed memory blocks don't have
VA to PA mapping created in kernel page table. Kmemleak scan on
removed memory blocks is causing page faults and leading to
kernel panic. So, Disable kmemleak scan on the removed memory
blocks.

Unable to handle kernel paging request at virtual address ffff80097b000000
pgd = ffff80096cb8f000
[ffff80097b000000] *pgd=0000000000000000
Internal error: Oops: 96000006 [#1] PREEMPT SMP
CPU: 2 PID: 285 Comm: kmemleak Not tainted 4.12.29-yocto-standard #2
Hardware name: LS1046A RDB Board (DT)
task: ffff80096db32f00 task.stack: ffff00000ad70000
PC is at scan_block+0x6c/0x180
LR is at scan_block+0x68/0x180
pc : [<ffff000008220044>] lr : [<ffff000008220040>] pstate: 800001c5
sp : ffff00000ad73d30
x29: ffff00000ad73d30 x28: 0000000000000000
x27: ffff000008f18000 x26: 00000000000001c0
x25: ffff0000090b5590 x24: ffff80096f122000
x23: ffff0000090b5000 x22: ffff000008f5a0e0
x21: ffff80097b000ff9 x20: ffff80097b001000
x19: ffff80097b000000 x18: 0000ffffc3fca58f
x17: 0000ffffb1cc26e0 x16: ffff000008127d78
x15: 00004544f8000000 x14: 002d0fa500000000
x13: 000000005bdac6ad x12: 0000000000000018
x11: 0000000027009ac0 x10: 00000000000007f0
x9 : ffff00000ad73ce0 x8 : ffff80096db33750
x7 : 0000000000600000 x6 : 0000000000000018
x5 : 0000000000000000 x4 : 0000000000000000
x3 : ffff00000ad73d90 x2 : ffff80096f122000
x1 : 0000000000208040 x0 : 0000000000000000
Process kmemleak (pid: 285, stack limit = 0xffff00000ad70000)
Call trace:
Exception stack(0xffff00000ad73bf0 to 0xffff00000ad73d30)
3be0: 0000000000000000 0000000000208040
3c00: ffff80096f122000 ffff00000ad73d90 0000000000000000 0000000000000000
3c20: 0000000000000018 0000000000600000 ffff80096db33750 ffff00000ad73ce0
3c40: 00000000000007f0 0000000027009ac0 0000000000000018 000000005bdac6ad
3c60: 002d0fa500000000 00004544f8000000 ffff000008127d78 0000ffffb1cc26e0
3c80: 0000ffffc3fca58f ffff80097b000000 ffff80097b001000 ffff80097b000ff9
3ca0: ffff000008f5a0e0 ffff0000090b5000 ffff80096f122000 ffff0000090b5590
3cc0: 00000000000001c0 ffff000008f18000 0000000000000000 ffff00000ad73d30
3ce0: ffff000008220040 ffff00000ad73d30 ffff000008220044 00000000800001c5
3d00: ffff0000090b55b0 00000000000001c0 ffffffffffffffff ffff00000821ffcc
3d20: ffff00000ad73d30 ffff000008220044
[<ffff000008220044>] scan_block+0x6c/0x180
[<ffff00000822023c>] scan_gray_list+0xe4/0x190
[<ffff0000082205f0>] kmemleak_scan+0x308/0x638
[<ffff0000082209ac>] kmemleak_scan_thread+0x8c/0xe0
[<ffff0000080c9ad8>] kthread+0x138/0x140
[<ffff0000080854a0>] ret_from_fork+0x10/0x18
Code: 910382d6 f9000fb4 97ffffd3 35000540 (f9400260)
---[ end trace d32f993a8921885a ]---
note: kmemleak[285] exited with preempt_count 2

Signed-off-by: Jiang Lu <lu.jiang@windriver.com>
---
 drivers/of/of_reserved_mem.c |    4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/drivers/of/of_reserved_mem.c b/drivers/of/of_reserved_mem.c
index d507c35..9b161e6 100644
--- a/drivers/of/of_reserved_mem.c
+++ b/drivers/of/of_reserved_mem.c
@@ -54,8 +54,10 @@ int __init __weak early_init_dt_alloc_reserved_memory_arch(phys_addr_t size,
 	}
 
 	*res_base = base;
-	if (nomap)
+	if (nomap) {
+		kmemleak_ignore_phys(base);
 		return memblock_remove(base, size);
+	}
 	return 0;
 }
 #else
-- 
1.7.9.5

